File: AES_PY/AES256ECB.py
#credit to tcitry on GitHub for base implementation
#Lightly edited by Alexander Korn, 11/24

import base64
import time
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.primitives import padding
from cryptography.hazmat.backends import default_backend
from django.utils.encoding import force_bytes, force_str

SECRET_KEY = "WordsMeaninglessAsWickedAndCruel"
#value = force_bytes("Danny Rensch and Levy Rozman are the best people I know. But really all I need to do is to provide a bunch of random text and characters, like this: ?*!!!*?. You weren't expecting a normal sentence, were you? It's a test and in ASCII I need 256 characters.")
value = force_bytes("Danny Rensch and")

backend = default_backend()
key = force_bytes(base64.urlsafe_b64encode(force_bytes(SECRET_KEY))[:32])


class Crypto:

    def __init__(self):
        self.encryptor = Cipher(algorithms.AES(key), modes.ECB(), backend).encryptor()
        self.decryptor = Cipher(algorithms.AES(key), modes.ECB(), backend).decryptor()

    def encrypt(self):
        #padder = padding.PKCS7(algorithms.AES(key).block_size).padder()
        #padded_data = padder.update(value) + padder.finalize()
        padded_data = value
        encrypted_text = self.encryptor.update(padded_data) #+ self.encryptor.finalize()
        return encrypted_text

    def decrypt(self, value):
        #padder = padding.PKCS7(algorithms.AES(key).block_size).unpadder()
        decrypted_data = self.decryptor.update(value)
        return decrypted_data
        #unpadded = padder.update(decrypted_data) + padder.finalize()
        #return unpadded


if __name__ == '__main__':
    print('>>>>>>>>>>>')
    crypto = Crypto()
    #text = force_str(base64.urlsafe_b64encode(crypto.encrypt()))
    enc_start = time.time()
    text = crypto.encrypt()
    enc_end = time.time()
    suptext = crypto.decrypt(base64.urlsafe_b64decode(force_str(base64.urlsafe_b64encode(text))))
    dec_end = time.time()
    print("CPU Execution Times")
    print("Elapsed Encryption: " + str(enc_end-enc_start))
    print("Elapsed Decryption: " + str(dec_end-enc_end))
    print("Elapsed Total: " + str(dec_end-enc_start))
    hextext = text.hex()
    print(hextext)
    print('<<<<<<<<<<<<<')
    #text = crypto.decrypt(base64.urlsafe_b64decode(force_str(base64.urlsafe_b64encode(text))))
    print(suptext)

File: AES_CPP/aes256ecb.cpp
#include <iostream>
#include <string>
#include <chrono>
#include <openssl/aes.h>
#include <openssl/evp.h>
#include <openssl/rand.h>
#include <openssl/bio.h>
#include <openssl/buffer.h>
#include <openssl/err.h>
#include <vector>
#include <iomanip>

#define SECRET_KEY "WordsMeaninglessAsWickedAndCruel"
//#define VALUE "Danny Rensch and"
#define VALUE "Danny Rensch and Levy Rozman are the best people I know. But really all I need to do is to provide a bunch of random text and characters, like this: ?*!!!*?. You weren't expecting a normal sentence, were you? It's a test and in ASCII I need 256 characters."

class Crypto {
private:
    EVP_CIPHER_CTX* encryptCtx;
    EVP_CIPHER_CTX* decryptCtx;
    std::vector<unsigned char> key;

public:
    Crypto() {
        encryptCtx = EVP_CIPHER_CTX_new();
        decryptCtx = EVP_CIPHER_CTX_new();

        // Generate key from SECRET_KEY
        std::string encoded_key = SECRET_KEY;
        encoded_key = encoded_key.substr(0, 32); // Ensure the key is 32 bytes long
        key.assign(encoded_key.begin(), encoded_key.end());

        // Initialize encryption context
        EVP_EncryptInit_ex(encryptCtx, EVP_aes_256_ecb(), nullptr, key.data(), nullptr);

        // Initialize decryption context
        EVP_DecryptInit_ex(decryptCtx, EVP_aes_256_ecb(), nullptr, key.data(), nullptr);
    }

    ~Crypto() {
        EVP_CIPHER_CTX_free(encryptCtx);
        EVP_CIPHER_CTX_free(decryptCtx);
    }

    std::vector<unsigned char> encrypt(const std::string& value) {
        std::vector<unsigned char> encrypted(value.size() + AES_BLOCK_SIZE);
        int len;
        EVP_EncryptUpdate(encryptCtx, encrypted.data(), &len, reinterpret_cast<const unsigned char*>(value.data()), value.size());
        encrypted.resize(len);
        return encrypted;
    }

    std::string decrypt(const std::vector<unsigned char>& encrypted_data) {
        std::vector<unsigned char> decrypted(encrypted_data.size());
        int len;
        EVP_DecryptUpdate(decryptCtx, decrypted.data(), &len, encrypted_data.data(), encrypted_data.size());
        decrypted.resize(len);
        return std::string(decrypted.begin(), decrypted.end());
    }
};

int main() {
    std::cout << ">>>>>>>>>>>>>>>>>>" << std::endl;

    Crypto crypto;
	//change code 2 lines down to SHORT_VALUE for 
    // Encryption
    std::chrono::high_resolution_clock::time_point enc_start = std::chrono::high_resolution_clock::now();
    std::vector<unsigned char> encrypted_data = crypto.encrypt(VALUE);
    std::chrono::high_resolution_clock::time_point enc_end = std::chrono::high_resolution_clock::now();

    // Decryption
    std::chrono::high_resolution_clock::time_point dec_start = std::chrono::high_resolution_clock::now();
    std::string decrypted_data = crypto.decrypt(encrypted_data);
    std::chrono::high_resolution_clock::time_point dec_end = std::chrono::high_resolution_clock::now();

    int enc_time = std::chrono::duration_cast<std::chrono::nanoseconds>(enc_end-enc_start).count();
    int dec_time = std::chrono::duration_cast<std::chrono::nanoseconds>(dec_end - dec_start).count();

    // Execution times
    std::cout << "CPU Execution Times" << std::endl;
    std::cout << "Elapsed Encryption: " << enc_time << " nanoseconds" << std::endl;
    //std::cout << "Elapsed Encryption: " << std::chrono::duration_cast<std::chrono::nanoseconds>(enc_time).count() << " nanoseconds" << std::endl;
    std::cout << "Elapsed Decryption: " << dec_time << " nanoseconds" << std::endl;
    std::cout << "Elapsed Total: " << (enc_time + dec_time) << " nanoseconds" << std::endl;

    // Print encrypted data as hex
    std::cout << "Hex Ciphertext: ";
    for (unsigned char byte : encrypted_data) {
        std::cout << std::hex << std::setw(2) << std::setfill('0') << (int)byte;
    }
    std::cout << std::endl;

    // Print decrypted data
    std::cout << "Decrypted Text: " << decrypted_data << std::endl;
    std::cout << "<<<<<<<<<<<<<<<<<<" << std::endl;

    return 0;
}
